README
NAME
    Task::Weaken - Ensure that a platform has weaken support

DESCRIPTION
    One recurring problem in modules that use Scalar::Util's "weaken"
    function is that it is not present in the pure-perl variant.

    While this isn't necesarily always a problem in a straight CPAN-based
    Perl environment, some operating system distributions only include the
    pure-Perl versions, don't include the XS version, and so weaken is then
    "missing" from the platform, despite passing a dependency on
    Scalar::Util successfully.

    Most notably this is RedHat Linux at time of writing, but other come and
    go and do the same thing, hence "recurring problem".

    The normal solution is to manually write tests in each distribution to
    ensure that "weaken" is available.

    This restores the functionality testing to a dependency you do once in
    your Makefile.PL, rather than something you have to write extra tests
    for each time you write a module.

    It should also help make the package auto-generators for the various
    operating systems play more nicely, because it introduces a dependency
    that they have to have a proper weaken in order to work.

  How this Task works
    Part of the problem seems to stem from the fact that some distributions
    continue to include modules even if they fail some of their tests.

    To get around that for this module, it will do a few dirty tricks.

    If Scalar::Util is not available at all, it will issue a normal
    dependency on the module. However, if Scalar::Util is relatively new (
    it is >= 1.14 ) and the module does not have weaken, the install will
    bail out altogether with a "platform does not support this module"
    error.

    This means that a distributor cannot just force the install, as there
    will be no Makefile at all to do the installation phase.

    This is to some degree lying, so it is somewhat debatable if this
    strategy should continue, and may be removed later.

    This distribution also contains tests to ensure that weaken is available
    using more normal methods.

    So if your module uses "weaken", you can just add the following to your
    Module::Install-based Makefile.PL (or equivalent).

      requires 'Task::Weaken';

AUTHOR
    Adam Kennedy <cpan@ali.as>, <http://ali.as/>

SEE ALSO
    Task, Scalar::Util, <http://ali.as/>

COPYRIGHT
    Copyright 2006 Adam Kennedy. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Task-Weaken

{{$NEXT}}

1.05      2017-12-21 01:10:56Z
        - switch packaging to Dist::Zilla
        - Require List::Util::XS in some cases (unreleased commit from 2012,
          Alexandr Ciornii)

1.04      2011-03-08
	- Updating author tests (ADAMK)
	- Updating to Module::Install (ADAMK)
        - Simple refactoring of Makefile.PL (CHORNY)

1.03      2009-06-12
	- Updating to flush out some old CPAN Testers results

1.02      2007-11-14
	- Updating to newer AUTOMATED_TESTING tests
	- Updating to fixed META.yml

1.01      2007-10-22
	- Updating to newer AUTOMATED_TESTING tests

1.00      2007-08-18
	- Adding more detail to the Makefile.PL, better error messages,
	  and several additional fallback options.

0.99      2006-05-25
	- Initial release


lib/Task/Weaken.pm
package Task::Weaken;

use 5.005;
use strict;

our $VERSION = '1.06';

1;

__END__

=pod

=head1 NAME

Task::Weaken - Ensure that a platform has weaken support

=head1 DESCRIPTION

One recurring problem in modules that use L<Scalar::Util>'s C<weaken>
function is that it is not present in the pure-perl variant.

While this isn't necessarily always a problem in a straight CPAN-based
Perl environment, some operating system distributions only include the
pure-Perl versions, don't include the XS version, and so weaken is
then "missing" from the platform, B<despite> passing a dependency on
L<Scalar::Util> successfully.

Most notably this is RedHat Linux at time of writing, but other come
and go and do the same thing, hence "recurring problem".

The normal solution is to manually write tests in each distribution
to ensure that C<weaken> is available.

This restores the functionality testing to a dependency you do once
in your F<Makefile.PL>, rather than something you have to write extra
tests for each time you write a module.

It should also help make the package auto-generators for the various
operating systems play more nicely, because it introduces a dependency
that they B<have> to have a proper weaken in order to work.

=head2 How this Task works

Part of the problem seems to stem from the fact that some distributions
continue to include modules even if they fail some of their tests.

To get around that for this module, it will do a few dirty tricks.

If L<Scalar::Util> is not available at all, it will issue a normal
dependency on the module. However, if L<Scalar::Util> is relatively
new ( it is E<gt>= 1.19 ) and the module does B<not> have weaken, the
install will bail out altogether with a long error encouraging the
user to seek support from their vendor (this problem happens most
often in vendor-packaged Perl versions).

This distribution also contains tests to ensure that weaken is
available using more normal methods.

So if your module uses C<weaken>, you can just add the following to
your L<Module::Install>-based F<Makefile.PL> (or equivalent).

  requires 'Task::Weaken' => 0;

=head1 SUPPORT

Bugs should be always be reported via the CPAN bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Task-Weaken>

For other issues,contact the author.

=head1 AUTHOR

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

=head1 SEE ALSO

L<Task>, L<Scalar::Util>, L<http://ali.as/>

=head1 COPYRIGHT

Copyright 2006 - 2011 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut


Makefile.PL
# Dear Distribution Packagers
#
# Hi there!
#
# Before we begin, I'm really really sorry for what you are about to
# encounter a bit lower down in this file.
#
# Trust me, I didn't want to WRITE it just as much as you are probably
# not going to want to READ it and try to work out how the hell it
# applies to your Linux/etc distribution.
#
# So lets try to shortcut the whole confusion thing and save us both
# some time.
#
# Some distributions have this problem where the Scalar::Util version
# in the Perl core gets built WITHOUT a C compiler available. When this happens,
# most of the functions in Scalar::Util that need the compiler are replaced
# with a Pure-Perl version that works identically...
#
# ...except for two function relating to "weak references", a reference
# that is intentionally not counted for the purposes of garbage collection.
#
# It's used to implement certain specific data structures, like circular
# references and parent links in tree structures, in a way that can safely
# be garbage collected and won't leak.
#
# So what you really should check FIRST, is that the Perl build in your
# distribution has the Scalar::Util::weaken function defined.
#
# If not, then your Perl build is probably broken, and your solution
# ultimately needs to be applied to Perl itself. Then the whole problem
# goes away.
#
# If Scalar::Util::weaken is defined in your core Perl distribution, then
# none of the edge cases described below apply, and everything will be
# just rosy.
#
# Best Regards
#
# Adam K

use inc::Module::Install 1.00;

all_from      'lib/Task/Weaken.pm';
test_requires 'Test::More' => '0.42';
test_requires 'File::Spec' => '0.80';
perl_version  '5.006';

check_scalar_util();

WriteAll;

sub check_scalar_util {
	# Try to load Scalar::Util
	eval {
		require Scalar::Util;
	};
	if ( $@ ) {
		# Scalar::Util not installed (really old Perl?)
		requires 'Scalar::Util' => '1.14';
		return;
	}

	# Is installed
	my $module_version = $Scalar::Util::VERSION;
	my $module_weaken  = !! defined &Scalar::Util::weaken;
	if ( $module_weaken ) {
		# Already have weaken support.
		# Make sure it is a new enough version
		requires 'Scalar::Util' => '1.14';
		return;
	}

	# Does NOT have weaken, so either a very old
	# Scalar::Util, or a broken one.
	if ( $module_version <= 1.01 ) {
		# Very old Scalar::Util, upgrade it
		requires 'Scalar::Util' => '1.14';
		return;
	}

	my $latest_scalar_util = 1.21; #does not support 5.5, but weak references were introduced in 5.6
	$latest_scalar_util = 1.19 if $] < 5.006;
	# will not work of course - CPAN automatically installs only latest version
        
	# Broken Scalar::Util
	# That is, it does NOT contain the weaken function
	# which means it was built without a compiler.
	unless ( can_cc() ) {
		# The user does not have a compiler
		# There is currently no way to resolve
		# this situation, but we should probably
		# try to upgrade Scalar::Util anyway, just
		# in case the author is able to come up
		# with a solution at some point in the
		# future.
		message1();
		requires 'List::Util::XS' => $latest_scalar_util;
		return;
	}

	# User should be capable of installing an
	# upgraded version.
	# Can we be certain they WILL do the upgrade?
	if ( $module_version < $latest_scalar_util ) {
		# We should be able to do a straight
		# forward upgrade of the module
		requires 'List::Util::XS' => $latest_scalar_util;
		return;
	}

	# They ALREADY have the newest version... and it is broken.
	# Very little we can do in this case, other than issue a
	# message to the user, and then add a high dependency in the
	# vague hope it does something.
	message2();
	requires 'List::Util::XS' => $latest_scalar_util;
	return;
}

sub message1 {
	print "\n\n\n";
	print "    ERROR:\n\n";
	print "    A CPAN module critically requires a function\n";
	print "    (Scalar::Util::weaken) that can only be provided by\n";
	print "    upgrading your Scalar::Util module to a new version,\n";
	print "    which will need a C compiler in order to install.\n\n";
	print "    Unfortunately, I can't seem to locate a C compiler on this\n";
	print "    computer.\n\n";
	print "    I'm going to try to continue anyway, but the most likely result\n";
	print "    is going to be an extremely noisy series of testing errors.\n\n";
	print "    If this happens, you will need to install a C compiler\n";
	print "    (such as gcc) and then try to install whatever it is you are\n";
	print "    installing again.\n\n";
	print "    During the second attempt, I should be able to find the C\n";
	print "    compiler and be able to build the needed function without\n";
	print "    having to bother you again.\n\n";
	print "    I'm going to wait for about a minute now in the hope you read\n";
	print "    this\n\n\n";
	sleep( 50 );
}

sub message2 {
	print "\n\n\n";
	print "    ERROR:\n\n";
	print "    A CPAN module critically requires a function\n";
	print "    (Scalar::Util::weaken) that should exist in your\n";
	print "    Scalar::Util module but doesn't.\n\n";
	print "    This probably happened because you are using a Perl\n";
	print "    provided by a binary package from a vendor, and this\n";
	print "    vendor has packaged Perl incorrectly.\n\n";
	print "    I have checked for a couple of potential workarounds\n";
	print "    but none of them appear to be usable in your\n";
	print "    situation.\n\n";
	print "    I will try a last-ditch option anyway, but the most\n";
	print "    likely result is a number of noisily failing tests\n\n";
	print "    If this happens, you will need to contact technical\n";
	print "    support for your vendor and report the broken Perl,\n";
	print "    so that they can repair it.\n\n";
	print "    Please refer them to the documentation for the\n";
	print "    'Task::Weaken' CPAN module, which explains the problem\n";
	print "    and how they can fix it.\n\n";
	print "    I'm going to wait for about a minute now so you have time\n";
	print "    to read this message\n\n\n";
	sleep( 50 );
}
